Analysis/src/Coda3EventDecoder.cc

```diff
@@ -0,0 +1,422 @@
+  assert(buffer);
+	
+	// General Event information
+  fEvtLength = buffer[0]+1;  // in longwords (4 bytes)
+  fEvtType = 0;
+	fEvtTag = 0;
+  fBankDataType = 0;
+
+	// Prep TBOBJ variables
+	tbank.Clear();
+	tsEvType = 0;
+  evt_time = 0;
+	trigger_bits = 0;
+	block_size = 0;
+		
+	// Start Filling Data
+	fEvtTag     = (buffer[1] & 0xffff0000) >> 16;
+	fBankDataType = (buffer[1] & 0xff00) >> 8;
+	block_size  =	(buffer[1] & 0xff); 
+
+	if(block_size > 1) { QwWarning << "MultiBlock is not properly supported! block_size = " 
+											 				   << block_size << QwLog::endl; }
+	// Determine the event type by the call
+	fEvtType = InterpretBankTag(fEvtTag);		
+	fWordsSoFar = (2);
+	if(fEvtTag < 0xff00) { /* user event */ printUserEvent(buffer); }
+	else if(fControlEventFlag) {
+		fEvtNumber = 0;	ProcessControlEvent(fEvtType, &buffer[fWordsSoFar]);
+	}
+	else if(fPhysicsEventFlag) { 
+		ret = trigBankDecode( buffer );
+		if(ret != HED_OK) { trigBankErrorHandler( ret ); }
+		else { 
+			fEvtNumber = tbank.evtNum;
+			fWordsSoFar = 2 + tbank.len;
+		}
+	}
+	else { // Not a control event, user event, nor physics event. Not sure what it is
+      	 //  Arbitrarily set the event type to "fEvtTag".
+      	 //  The first two words have been examined.
+    QwWarning << "!!! Cannot determine the event type !!!" << QwLog::endl;
+		QwMessage << "Printing Event Buffer:";
+		QwMessage << "\n------------\n" << QwLog::endl;
+		for(size_t index = 0; fEvtLength; index++){
+ 			// TODO: // what if fEvtLength is gibberish because the event is gibberish?
+			QwMessage << "\t" << buffer[index];
+			if(index % 4 == 0){ QwMessage << QwLog::endl; }
+		}	
+		QwMessage << "\n------------\n" << QwLog::endl;
+
+ 		fEvtType = fEvtTag;	fEvtNumber = 0;
+	}
+
+	fFragLength = fEvtLength - fWordsSoFar;	
+  QwDebug << Form("buffer[0-1] 0x%x 0x%x ; ", buffer[0], buffer[1]);
+	PrintDecoderInfo(QwDebug);
+
+	return CODA_OK;
+}
+
+Bool_t Coda3EventDecoder::IsPhysicsEvent()
+{
+	return (fEvtType <= MAX_PHYS_EVTYPE);
+}
+
+//_____________________________________________________________________________
+UInt_t Coda3EventDecoder::InterpretBankTag( UInt_t tag )
+{
+  UInt_t evtyp{};
+  if( tag >= 0xff00 ) { // CODA Reserved bank type
+    switch( tag ) {
+      case 0xffd1:
+        evtyp = kPRESTART_EVENT; // MQwControlEvent (protected -> needs to be made public 
+        														// and function needs to be a QwEventBuffer Member)
+        // TODO:
+        // Should I call ProcessControlEvent here?
+        // (See discussion in VEventDecoder's inheritance)
+				fControlEventFlag = kTRUE;
+        break;
+      case 0xffd2:
+         evtyp = kGO_EVENT; // MQwControlEvent (protected -> needs to be made public 
+        														// and function needs to be a QwEventBuffer Member)
+        // TODO:
+        // Should I call ProcessControlEvent here?
+        // (See discussion in VEventDecoder's inheritance)
+				fControlEventFlag = kTRUE;
```

Comment:
// TODO:
        // Should I call ProcessControlEvent here?
        // (See discussion in VEventDecoder's inheritance)